File: apps/api/drift_api/tasks.py
from __future__ import annotations

import json
from typing import Dict, Optional

from drift_api.config import ApiConfig


class CloudTasksEnqueuer:
    def __init__(self, config: ApiConfig, tasks_client=None) -> None:
        self._config = config
        self._tasks_v2 = None
        if tasks_client is None:
            from google.cloud import tasks_v2

            tasks_client = tasks_v2.CloudTasksClient()
            self._tasks_v2 = tasks_v2
        self._client = tasks_client

    def enqueue_analysis_task(self, payload: Dict[str, object]) -> Optional[str]:
        self._config.require_tasks_config()
        parent = self._client.queue_path(
            self._config.tasks_project_id,
            self._config.tasks_location,
            self._config.tasks_queue,
        )
        http_request = {
            "http_method": self._http_post_method(),
            "url": self._config.worker_url,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        }
        if self._config.tasks_service_account_email:
            http_request["oidc_token"] = {
                "service_account_email": self._config.tasks_service_account_email,
            }

        response = self._client.create_task(
            request={
                "parent": parent,
                "task": {"http_request": http_request},
            }
        )
        return getattr(response, "name", None)

    def _http_post_method(self):
        if self._tasks_v2 is None:
            return "POST"
        return self._tasks_v2.HttpMethod.POST

File: apps/api/src/tasks.ts
import { CloudTasksClient } from "@google-cloud/tasks";
import { ApiConfig } from "./config";

export type AnalysisTaskPayload = {
  jobId: string;
  sourceArchiveUri: string;
  documentUris: string[];
  resultsPrefix: string;
  projectName?: string;
  requestedBy?: string;
};

type TasksClient = {
  queuePath(project: string, location: string, queue: string): string;
  createTask(request: {
    parent: string;
    task: {
      httpRequest: {
        httpMethod: "POST";
        url: string;
        headers: Record<string, string>;
        body: Buffer;
        oidcToken?: {
          serviceAccountEmail: string;
        };
      };
    };
  }): Promise<[{ name?: string } | undefined]>;
};

export interface TaskEnqueuer {
  enqueueAnalysisTask(payload: AnalysisTaskPayload): Promise<string | null>;
}

export class CloudTasksEnqueuer implements TaskEnqueuer {
  private readonly config: ApiConfig;
  private readonly client: TasksClient;

  constructor(config: ApiConfig, tasksClient: TasksClient | null = null) {
    this.config = config;
    this.client = tasksClient ?? (new CloudTasksClient() as unknown as TasksClient);
  }

  async enqueueAnalysisTask(payload: AnalysisTaskPayload): Promise<string | null> {
    this.config.requireTasksConfig();
    const parent = this.client.queuePath(
      this.config.tasksProjectId!,
      this.config.tasksLocation!,
      this.config.tasksQueue!,
    );
    const httpRequest: {
      httpMethod: "POST";
      url: string;
      headers: Record<string, string>;
      body: Buffer;
      oidcToken?: { serviceAccountEmail: string };
    } = {
      httpMethod: "POST",
      url: this.config.workerUrl!,
      headers: { "Content-Type": "application/json" },
      body: Buffer.from(JSON.stringify(payload), "utf8"),
    };
    if (this.config.tasksServiceAccountEmail) {
      httpRequest.oidcToken = {
        serviceAccountEmail: this.config.tasksServiceAccountEmail,
      };
    }

    const [response] = await this.client.createTask({
      parent,
      task: { httpRequest },
    });
    return response?.name || null;
  }
}
